### PR TITLE
Emit disk spill and merge buffer utilisation metrics for GroupBy queries

### PR DESCRIPTION
**Pull request by findingrish**
This change is to emit following metrics as part of `GroupByStatsMonitor` monitor,
`mergeBuffer/used` -> Number of merge buffers used. 
`mergeBuffer/acquisitionTimeNs` -> Total time required to acquire merge buffer. 
`mergeBuffer/acquisition` -> Number of queries that acquired a batch of merge buffers. 
`groupBy/spilledQueries` -> Number of queries that spilled onto the disk. 
`groupBy/spilledBytes`-> Spilled bytes on the disk. 
`groupBy/mergeDictionarySize` -> Size of the merging dictionary. 

Testing 
1. Added UT to verify that `GroupByStatsMonitor` is reporting the new metrics. 
2. Added UT for `GroupByStatsProvider`.
3. Tested metric emission in a local druid cluster for a group by query. 

``` 
{"feed":"metrics","metric":"mergeBuffer/usedCount","service":"druid/historical","host":"localhost:8083","version":"32.0.0-SNAPSHOT","value":0,"timestamp":"2024-10-24T13:09:48.884Z"}
{"feed":"metrics","metric":"mergeBuffer/acquisitionCount","service":"druid/historical","host":"localhost:8083","version":"32.0.0-SNAPSHOT","value":1,"timestamp":"2024-10-24T13:09:48.884Z"} 
{"feed":"metrics","metric":"mergeBuffer/acquisitionTimeNs","service":"druid/historical","host":"localhost:8083","version":"32.0.0-SNAPSHOT","value":67541,"timestamp":"2024-10-24T13:09:48.884Z"} 
{"feed":"metrics","metric":"groupBy/spilledQueries","service":"druid/historical","host":"localhost:8083","version":"32.0.0-SNAPSHOT","value":1,"timestamp":"2024-10-24T13:09:48.884Z"}
{"feed":"metrics","metric":"groupBy/spilledBytes","service":"druid/historical","host":"localhost:8083","version":"32.0.0-SNAPSHOT","value":367977,"timestamp":"2024-10-24T13:09:48.884Z"}
{"feed":"metrics","metric":"groupBy/mergeDictionarySize","service":"druid/historical","host":"localhost:8083","version":"32.0.0-SNAPSHOT","value":296,"timestamp":"2024-10-24T13:09:48.884Z"}
{"feed":"metrics","metric":"mergeBuffer/pendingRequests","service":"druid/historical","host":"localhost:8083","version":"32.0.0-SNAPSHOT","value":0,"timestamp":"2024-10-24T13:09:48.883Z"}

{"feed":"metrics","metric":"mergeBuffer/pendingRequests","service":"druid/broker","host":"localhost:8082","version":"32.0.0-SNAPSHOT","value":0,"timestamp":"2024-10-24T13:08:49.711Z"}
{"feed":"metrics","metric":"mergeBuffer/usedCount","service":"druid/broker","host":"localhost:8082","version":"32.0.0-SNAPSHOT","value":0,"timestamp":"2024-10-24T13:08:49.711Z"}
{"feed":"metrics","metric":"mergeBuffer/acquisitionCount","service":"druid/broker","host":"localhost:8082","version":"32.0.0-SNAPSHOT","value":1,"timestamp":"2024-10-24T13:08:49.711Z"}
{"feed":"metrics","metric":"mergeBuffer/acquisitionTimeNs","service":"druid/broker","host":"localhost:8082","version":"32.0.0-SNAPSHOT","value":2545250,"timestamp":"2024-10-24T13:08:49.711Z"}
{"feed":"metrics","metric":"groupBy/mergeDictionarySize","service":"druid/broker","host":"localhost:8082","version":"32.0.0-SNAPSHOT","value":20190,"timestamp":"2024-10-24T13:08:49.712Z"}
```

### Release Notes 

Going forward `GroupByStatsMonitor` monitor should be used for emitting `mergeBuffer/pendingRequests` metric instead of `QueryCountStatsMonitor`. In a subsequent release this metric will not be emitted from `QueryCountStatsMonitor`.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  {
+    return blockingPool.getUsedBufferCount();
+  }
+
+  public void registerTemporaryStorage(LimitedTemporaryStorage temporaryStorage)
+  {
+    temporaryStorages.add(temporaryStorage);
+  }
+
+  public long getSpilledBytes()
```
[the remark this answers is not in the document]
Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:69]: Regardless of this, I think that there should be metric (either this one after repurposing or a new one) which indicates the total size of the spilled data, per query. This will allow the admins to estimate whether the queries need a larger merge buffer and by how much. WDYT?

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ */
+@LazySingleton
+public class GroupByStatsProvider
+{
+  private final AtomicLong resourceAcquisitionTimeNs = new AtomicLong(0);
+  private final AtomicLong resourceAcquisitionCount = new AtomicLong(0);
+
+  private final ConcurrentLinkedQueue<LimitedTemporaryStorage> temporaryStorages;
+
+  @Inject
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:41]: nit: We shouldn't need this

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Collects stats for group by queries like used merged buffer count, spilled bytes and group by resource acquisition time.
+ */
+@LazySingleton
+public class GroupByStatsProvider
+{
+  private final AtomicLong resourceAcquisitionTimeNs = new AtomicLong(0);
+  private final AtomicLong resourceAcquisitionCount = new AtomicLong(0);
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:37]:
```suggestion
  private final AtomicLong mergeBufferAcquisitionTimeNs = new AtomicLong(0);
  private final AtomicLong mergeBufferAcquisitionCount = new AtomicLong(0);
```

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    this.temporaryStorages = new ConcurrentLinkedQueue<>();
+  }
+
+  public synchronized void groupByResourceAcquisitionTimeNs(long delayNs)
+  {
+    resourceAcquisitionTimeNs.addAndGet(delayNs);
+    resourceAcquisitionCount.incrementAndGet();
+  }
+
+  public synchronized long getAndResetGroupByResourceAcquisitionStats()
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:53]: This doesn't seem correct. We are reporting the average as a metric. A better way would be to report the sum, as well as the count. This will allow a better weighted-average than directly reporting the average from the monitor. While we are amortizing a lot of metrics by emitting it from the monitor, I think it's better to report count and sum separately.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java (hunk 3 above)
Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:37]: Making them atomic seems redundant given that both need to be updated together, and guaranteeing atomicity over a single variable is not useful. The current code does not rely on the `AtomicLong` part of the variables since they are already guarded by the provider's lock.

File: docs/configuration/index.md [hunk 5]
```diff
@@ -401,7 +401,7 @@ Metric monitoring is an essential part of Druid operations. The following monito
 |`org.apache.druid.java.util.metrics.CgroupV2MemoryMonitor`| **EXPERIMENTAL** Reports memory usage from `memory.current` and `memory.max` files. Only applicable to `cgroupv2`.|
 |`org.apache.druid.server.metrics.HistoricalMetricsMonitor`|Reports statistics on Historical services. Available only on Historical services.|
 |`org.apache.druid.server.metrics.SegmentStatsMonitor` | **EXPERIMENTAL** Reports statistics about segments on Historical services. Available only on Historical services. Not to be used when lazy loading is configured.|
-|`org.apache.druid.server.metrics.QueryCountStatsMonitor`|Reports how many queries have been successful/failed/interrupted.|
+|`org.apache.druid.server.metrics.QueryCountStatsMonitor`|Reports how many queries have been successful/failed/interrupted. It also reports stats for group by queries.|
```
Comment (LakshSingla) [inline docs/configuration/index.md:404]:
```suggestion
|`org.apache.druid.server.metrics.QueryCountStatsMonitor`|Reports how many queries have been successful/failed/interrupted. It also reports statistics for the group by queries.|
```

File: processing/src/main/java/org/apache/druid/collections/BlockingPool.java [hunk 6]
```diff
@@ -49,4 +49,9 @@ public interface BlockingPool<T>
    * @return count of pending requests
    */
   long getPendingRequests();
+
+  /**
+   * @return number of used buffers from the pool
+   */
+  long getUsedBufferCount();
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/collections/BlockingPool.java:56]: BlockingPool is generic. It shouldn't reference buffers. 
```suggestion
  long getUsedResourcesCount();
```

File: processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java [hunk 7]
```diff
@@ -119,7 +119,7 @@ public List<ReferenceCountingResourceHolder<T>> takeBatch(final int elementNum)
       throw new RuntimeException(e);
     }
     finally {
-      pendingRequests.incrementAndGet();
+      pendingRequests.decrementAndGet();
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java:122]: nice catch 💯

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    long average = resourceAcquisitionCount.get() != 0 ?
+                   (resourceAcquisitionTimeNs.get() / resourceAcquisitionCount.get()) : 0;
+
+    resourceAcquisitionTimeNs.set(0);
+    resourceAcquisitionCount.set(0);
+
+    return average;
+  }
+
+  public void registerTemporaryStorage(LimitedTemporaryStorage temporaryStorage)
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:64]: How often does this method need to be called per group by query, and what does this rely upon? I feel that there's an inbuilt expense with tracking the spilled bytes this way. I think it is still fine, but I wonder if a more efficient way would be as follows:

1. Have a running counter of the size occupied 
2. LimitedTemporaryStorage implements closeable. Modify wherever the LimitedTemporaryStorage is getting closed to subtract the value of the storage from the counter. There can be a set inclusion check where a hashset of the opened temporary storages being tracked by the running counter are stored, and the code confirms that the removed temporary storage was indeed being tracked by the counter. 

Note: My approach seems more convoluted, so if there isn't much performance downside of the current version, I think its fine as is.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (findingrish) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:69]: > Would the code double-count the stored bytes? Perhaps that is the intended behavior.

Yes, that is the intention. It basically is reporting the amount of bytes spilled when the metric is emitted. 

> Regardless of this, I think that there should be metric (either this one after repurposing or a new one) which indicates the total size of the spilled data, per query. This will allow the admins to estimate whether the queries need a larger merge buffer and by how much. 

Makes sense, but this would be reported at the end of every query?

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java (hunk 4 above)
Comment (findingrish) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:53]: I couldn't find a way wherein I could report the sum and count together (so that any metric aggregation platform could automatically compute the averages). 
Are you suggesting to report two different metrics `groupBy/acquisitionTime` & `groupBy/acquisitionCount`?

File: docs/configuration/index.md (hunk 5 above)
Comment (findingrish) [reply docs/configuration/index.md:404]: I don't have strong opinions either, the only reason I merged them was an existing mergeBuffer metric in this monitor. I think a separate monitor would be better due to the added flexibility.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java (hunk 8 above)
Comment (findingrish) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:64]: >there's an inbuilt expense with tracking the spilled bytes this way

What is the expense you are referring to? The temporary storage object is added to a queue and cost of adding an element should be in most cases should be O(1). 

>LimitedTemporaryStorage implements closeable. Modify wherever the LimitedTemporaryStorage is getting closed to subtract the value of the storage from the counter. There can be a set inclusion check where a hashset of the opened temporary storages being tracked by the running counter are stored, and the code confirms that the removed temporary storage was indeed being tracked by the counter.

This approach also requires keeping track of the temporary storages thus incurring similar overhead as the existing implementation. Additionally, it requires incrementing the "running counter" whenever bytes are written out to the disk.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:69]: Yes, the query can update the monitor at the end

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java (hunk 8 above)
Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:64]: > What is the expense you are referring to

Iterating while calculating the stored bytes.

File: server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java [hunk 9]
```diff
@@ -72,9 +75,11 @@ public boolean doMonitor(ServiceEmitter emitter)
       }
     }
 
-    long pendingQueries = this.mergeBufferPool.getPendingRequests();
-    emitter.emit(builder.setMetric("mergeBuffer/pendingRequests", pendingQueries));
+    if (emitMergeBufferPendingRequests) {
+      long pendingQueries = this.mergeBufferPool.getPendingRequests();
+      emitter.emit(builder.setMetric("mergeBuffer/pendingRequests", pendingQueries));
+    }
+
```
Comment (LakshSingla) [inline server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java:82]: Why is this required? Isn't it cleaner to not emit the metric from the GroupByStatsMonitor at all and keep it as is?

File: server/src/test/java/org/apache/druid/server/metrics/GroupByStatsMonitorTest.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  @Test(timeout = 2_000L)
+  public void testMonitoringMergeBuffer_pendingRequests()
+  {
+    executorService.submit(() -> {
+      mergeBufferPool.takeBatch(10);
+    });
+
+    int count = 0;
+    try {
+      // wait at most 10 secs for the executor thread to block
```
Comment (LakshSingla) [inline server/src/test/java/org/apache/druid/server/metrics/GroupByStatsMonitorTest.java:130]: Is the "10second" number stale?

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 11]
```diff
@@ -13682,4 +13755,16 @@ private void cannotVectorize()
       expectedException.expectMessage("Cannot vectorize!");
     }
   }
+
+  private void verifyMetrics(long queries, boolean skipMergeDictionary)
```
Comment (LakshSingla) [inline processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:13759]: What is skipMergeDictionary and when is it required?

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 12]
```diff
@@ -12383,6 +12448,10 @@ public void testGroupByOnNullableDoubleNoLimitPushdown()
 
     Iterable<ResultRow> results = GroupByQueryRunnerTestHelper.runQuery(factory, runner, query);
     TestHelper.assertExpectedObjects(expectedResults, results, "groupBy");
+
+    if (config.equals(V2_SMALL_BUFFER_CONFIG)) {
+      verifyMetrics(1, true);
+    }
```
Comment (LakshSingla) [inline processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:12454]: Why are the assertions being done on a single config?

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+    PerQueryStats container = perQueryStats.remove(resourceId);
+    aggregateStatsContainer.addQueryStats(container);
+  }
+
+  public synchronized AggregateStats getStatsSince()
+  {
+    return aggregateStatsContainer.reset();
+  }
+
+  public static class AggregateStats
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:66]: Why are some stats AggregateStats, and some PerQuery? Wouldn't the monitor aggregate them all?

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+import org.apache.druid.query.QueryResourceId;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Metrics collector for groupBy queries like spilled bytes, merge buffer acquistion time, dictionary size.
+ */
+@LazySingleton
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:32]: Should it be registered in a Module if this is a LazySingleton? (I am not very familiar, just curious if this is working as expected)

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+  }
+
+  public synchronized void closeQuery(QueryResourceId resourceId)
+  {
+    if (resourceId == null || !perQueryStats.containsKey(resourceId)) {
+      return;
+    }
+    PerQueryStats container = perQueryStats.remove(resourceId);
+    aggregateStatsContainer.addQueryStats(container);
+  }
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:59]: This is not intuitive. The callers won't remember calling `closeQuery()`. Is it possible to mold it as a Closer/Closable? PerQueryStats can extend Closeable, and the close can do that. 

It is still not good since it doesn't seem cool to close "Stats" in first place.  Perhaps renaming `PerQueryStats` to `PerQueryStatsContainer` would also help.

File: docs/operations/metrics.md [hunk 16]
```diff
@@ -103,7 +109,13 @@ Most metric values reset each emission period, as specified in `druid.monitoring
...
+|`mergeBuffer/usedCount`|Number of merge buffers used from the merge buffer pool.|This metric is only available if the `GroupByStatsMonitor` module is included.|Depends on the number of groupBy queries needing merge buffers.|
+|`mergeBuffer/acquisitionTimeNs`|Total time in nanoseconds to acquire merge buffer for groupBy queries.|This metric is only available if the `GroupByStatsMonitor` module is included.|Should be as low as possible.|
+|`mergeBuffer/acquisitionCount`|Number of times groupBy queries acquired merge buffers.|This metric is only available if the `GroupByStatsMonitor` module is included.|Depends on the number of groupBy queries needing merge buffers.|
+|`groupBy/spilledQueries`|Number of groupBy queries that have spilled onto the disk.|This metric is only available if the `GroupByStatsMonitor` module is included.|Varies|
+|`groupBy/spilledBytes`|Number of bytes spilled on the disk by the groupBy queries.|This metric is only available if the `GroupByStatsMonitor` module is included.|Varies|
+|`groupBy/mergeDictionarySize`|Size of on-heap merge dictionary in bytes.|This metric is only available if the `GroupByStatsMonitor` module is included.|Varies|
```
[the remark this answers is not in the document]
Comment (findingrish) [reply docs/operations/metrics.md:118]: Moved them to a new section.

File: server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java (hunk 9 above)
Comment (findingrish) [reply server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java:82]: Ideally, all the merge buffer metrics should be reported from the new monitor. This change is to ensure backward compatibility.

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java (hunk 11 above)
Comment (findingrish) [reply processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:13759]: Renamed, this is basically to skip verifying the merge dictionary size metric, since some of the queries aggregate on non-string dimensions.

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java (hunk 12 above)
Comment (findingrish) [reply processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:12454]: This is the most suited config to verify all the metrics. For example, in config with larger buffer size it is not possible to verify disk spill metrics. I am moved this check into the method and renamed the method.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java (hunk 14 above)
Comment (findingrish) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:32]: Yes, it works as expected. Guice should automatically detect classes marked with this annotation and bind them in singleton scope.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java (hunk 15 above)
Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:59]: >  it would require to be explicitly registered with the query closer

Yes. Implementing the interface makes it easier for the developer to call the closer. I have had experience in the past where an implicit contract was not apparent which led to resource leak

File: server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java (hunk 9 above)
Comment (LakshSingla) [reply server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java:82]: But we don't have any plans of removing this code right? Which makes me wonder if the additional complexity is worth it. The situation is that we have had a metric in past, which was emitted without a monitor, and we are not planning on modifying the behavior (LMK if this part is not true, which will make the comment null & void). So it doesn't seem correct that we are emitting this from the monitor as well for correctness given that we also have weird artefacts of this metric in QueryCountStatsMonitor which would not be removed.

File: server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
+
+    emitter.emit(builder.setMetric("mergeBuffer/pendingRequests", mergeBufferPool.getPendingRequests()));
+
+    emitter.emit(builder.setMetric("mergeBuffer/usedCount", mergeBufferPool.getUsedResourcesCount()));
+
+    GroupByStatsProvider.AggregateStats statsContainer = groupByStatsProvider.getStatsSince();
+
+    if (statsContainer.getMergeBufferAcquisitionCount() > 0) {
+      emitter.emit(builder.setMetric("mergeBuffer/acquisitionCount", statsContainer.getMergeBufferAcquisitionCount()));
```
Comment (LakshSingla) [inline server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java:59]: usedCount and acquisitionCount are difficult to disambiguate. Can we rename them?

**Review by LakshSingla — APPROVED** [review]

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java (hunk 14 above)
Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java:32]: Thanks for confirming!

File: server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java (hunk 17 above)
Comment (findingrish) [reply server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java]: Renamed `mergeBuffer/usedCount` -> `mergeBuffer/used`, `mergeBuffer/acquistionCount` -> `mergeBuffer/queries`.

LMK if this is easier to understand?

File: server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java (hunk 9 above)
Comment (findingrish) [reply server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java:82]: > The situation is that we have had a metric in past, which was emitted without a monitor, and we are not planning on modifying the behavior 

This isn't right, the metric was emitted from a monitor earlier (QueryCountStatsMonitor.java). 

We can remove the metric from the old monitor and the corresponding code in the next release?

File: docs/operations/metrics.md [hunk 18]
```diff
@@ -86,6 +85,13 @@ Most metric values reset each emission period, as specified in `druid.monitoring
...
 |`query/rowLimit/exceeded/count`|Number of queries whose inlined subquery results exceeded the given row limit|This metric is only available if the `SubqueryCountStatsMonitor` module is included.| |
 |`query/byteLimit/exceeded/count`|Number of queries whose inlined subquery results exceeded the given byte limit|This metric is only available if the `SubqueryCountStatsMonitor` module is included.| |
+|`mergeBuffer/pendingRequests`|Number of requests waiting to acquire a batch of buffers from the merge buffer pool.|This metric is only available if the `GroupByStatsMonitor` module is included.|Should be 0.|
+|`mergeBuffer/usedCount`|Number of merge buffers used from the merge buffer pool.|This metric is only available if the `GroupByStatsMonitor` module is included.|Depends on the number of groupBy queries needing merge buffers.|
+|`mergeBuffer/acquisitionCount`|Number of times groupBy queries acquired merge buffers.|This metric is only available if the `GroupByStatsMonitor` module is included.|Depends on the number of groupBy queries needing merge buffers.|
```
[the remark this answers is not in the document]
Comment (findingrish) [reply docs/operations/metrics.md]: It is cumulative, I have renamed it to `mergeBuffer/queries`.

**Comment by LakshSingla** [comment]
Waiting on a couple of changes. Will merge this post that.

**Comment by LakshSingla** [comment]
The minor change was to use query id instead of resource reservation id in the logic, but that is not a user facing thing and can be picked up in subsequent patches. I have merged this.

File: server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  @Override
+  public boolean doMonitor(ServiceEmitter emitter)
+  {
+    final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
+
+    emitter.emit(builder.setMetric("mergeBuffer/pendingRequests", mergeBufferPool.getPendingRequests()));
+
+    emitter.emit(builder.setMetric("mergeBuffer/used", mergeBufferPool.getUsedResourcesCount()));
```
Comment (TessaIO) [inline server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java:54]: @LakshSingla @findingrish Is there a reason why this metric isn’t included in the per-query stats? I think it would be helpful to report it for each query individually.

Right now, the emission of this metric depends on the emission interval. For example, if the interval is 1 minute and a query runs in the middle of that cycle—using and then freeing the merge buffers before the minute ends—we won’t see that the buffers were ever used. So, we could miss important usage information simply because the timing doesn’t align with the metric emission.

Please let me know if I’ve misunderstood how this works.

Comment (TessaIO) [reply server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java:54]: I enabled the metric on a test cluster, and I found that, despite running many GroupBy queries,`mergeBuffer/used ` always equals 0

Comment (TessaIO) [reply server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java:54]: cc @gianm

Comment (gianm) [reply server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java:54]: This metric is meant to be a point-in-time measurement, designed to help diagnose situations where merge buffer usage is consistently very high, and causes the pool to be continuously exhausted.

There is another metric `mergeBuffer/queries` that is the number of queries made that used merge buffers. This one might be more what you want, if you're wanting to see per query metrics.

Comment (TessaIO) [reply server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java:54]: @gianm `mergeBuffer/queries` defines how many queries acquired some merge buffers. But we can't get exact numbers, some queries acquire 1 merge buffer, and others acquire 2 merge buffers.

Comment (gianm) [reply server/src/main/java/org/apache/druid/server/metrics/GroupByStatsMonitor.java:54]: Yeah, that's a good point. IMO it would make sense to add a new metric like `mergeBuffer/count` which is the total number of merge buffers allocated. Similar to `mergeBuffer/queries`, but could increment by more than 1 per query.
